### PR TITLE
[FLINK-1909] [streaming] Type handling refactor for sources + scala api

### PR DESCRIPTION
**Pull request by gyfora**
This PR reworks the way sources handle type extraction, allowing for missing types (with returns) and also fixes some issues around type handling for the streaming Scala api.

### CODE REVIEW

**Comment by mbalassi** [comment]
LGTM.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 1]
```diff
@@ -632,15 +608,18 @@ public void registerType(Class<?> type) {
...
-				outTypeInfo = ((GenericSourceFunction<OUT>) function).getType();
-			} else {
+		if (function instanceof GenericSourceFunction) {
+			outTypeInfo = ((GenericSourceFunction<OUT>) function).getType();
+		} else {
+			try {
 				outTypeInfo = TypeExtractor.createTypeInfo(SourceFunction.class,
 						function.getClass(), 0, null, null);
+			} catch (InvalidTypesException e) {
+				outTypeInfo = (TypeInformation<OUT>) new MissingTypeInfo("Custom source", e);
```
Comment (rmetzger) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: Why did you change this?
I suspect the exception will now be thrown somewhere else?

Comment (gyfora) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: This is the same mechanism that all flink operators use to allow missing types that can be filled with .returns(..) afterwards.

Comment (rmetzger) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: Okay, I see
